File: classes/timeMeasures.py
"""Defining a time interval and how to calculate with it"""

from itertools import zip_longest

#6:2
HOURS_IN_DAY = 24
"""The day is broken up into 24 hours. """

class TimeInterval:
    """Used for the lenght of a month, year, etc."""
    parts_in_hour = 1   
    # A class variable. Could conceviably be set to any value.

    def __init__(self, days=0, hours=0, parts=0, *, parts_in_hour=None):
        if parts_in_hour:   
        # 0 would cause a division by zero, so rule that out too.
            TimeInterval.parts_in_hour = parts_in_hour
        self.days = days
        self.hours = hours
        self.parts = parts
        self.reduce()

    def reduce(self):
        """Reduce the number of parts to less than an hour, etc.
        
        Reduces the number of parts to less than an hour and the hours 
        to less than 24, adding the whole hours and whole days. Converts 
        fractional parts of days and hours to hours and parts. Does not 
        affect the whole day count."""
        #6:9
        # Convert fractional days into hours
        self.hours += (self.days % 1) * HOURS_IN_DAY
        self.days = int(self.days // 1)
        # Convert fractional hours into parts
        self.parts += (self.hours % 1) * TimeInterval.parts_in_hour
        self.hours = int(self.hours // 1)
        # Set the parts
        self.reduce_parts()
        # Carry the whole hours, then round the remaining parts.
        # (This also works for negetive inputs.)
        self.hours += self.parts // TimeInterval.parts_in_hour
        self.parts %= TimeInterval.parts_in_hour        
        # Carry the whole days, then round the remaining hours.
        self.days += self.hours // HOURS_IN_DAY
        self.hours %= HOURS_IN_DAY

    def reduce_parts(self):
        # Fractional parts will be ignored
        # untill the subclass FineTimeInterval
        self.parts = int(self.parts // 1)

    # Iteration function
    def __iter__(self) -> int:
        yield from [self.days, self.hours, self.parts]

    # Hash function
    def __hash__(self) -> int:
        return hash((self.days, self.hours, self.parts, self.parts_in_hour))
    # String functions
    def __str__(self) -> str:
        return f"{self.days} {self.hours:2} {self.parts:4}"
    def __repr__(self) -> str:
        return (f"{self.__class__.__name__}({self.days}, {self.hours}, "
        f"{self.parts}, parts_in_hour={self.parts_in_hour})")
        
    # The following methods work unchanged in a four item subclass.
    # absolute value
    def __abs__(self):
        return self * -1 if self < type(self)(0) else self

    # comparison functions
    def __eq__(self, other) -> bool:
        for x, y in zip_longest(self, other, fillvalue= 0):
            if x == y:
                continue
            else:
                return False
        return True

    def __gt__(self, other) -> bool:
        for x, y in zip_longest(self, other, fillvalue= 0):
            if   x >  y:
                return True
            elif x == y:
                continue
            else: # x < y
                return False
        # if x == y for all
        return False

    def __ge__(self, other) -> bool:
        for x, y in zip_longest(self, other, fillvalue= 0):
            if   x >  y:
                return True
            elif x == y:
                continue            
            else: # x < y
                return False
        # if x == y for all
        return True

    def __lt__(self, other) -> bool:
        for x, y in zip_longest(self, other, fillvalue= 0):
            if   x <  y:
                return True
            elif x == y:
                continue          
            else: # x > y
                return False
        # if x == y for all
        return False

    # math functions
    def __add__(self, addend):
        return type(self)(*[x + y
         for x, y in zip_longest(self, addend, fillvalue= 0)])
    def __radd__(self, addend):
        return type(self)(*[x + y
         for x, y in zip_longest(addend, self, fillvalue= 0)])
    def __sub__(self, subtrahend):
        return type(self)(*[x - y 
         for x, y in zip_longest(self, subtrahend, fillvalue= 0)])
    def __mul__(self, factor):
        return type(self)(*[x * factor for x in self])
    def __rmul__(self, factor):
        return type(self)(*[x * factor for x in self])
    def __floordiv__(self, divisor):
        return type(self)(*[x / divisor for x in self])
    def __truediv__(self, divisor):
        return FineTimeInterval(*[x / divisor for x in self],
         moments_in_part= divisor)

class TimeInWeek (TimeInterval):
    """A time of week, or the offset of a time of week."""
    def reduce(self):
        """Sets the day to a day of the week to 1-7."""
        super().reduce()
        self.days %= 7
        # We want Shabbos to be appear as 7, even though its mod is 0.
        if self.days == 0:
            self.days = 7

class FineTimeInterval(TimeInterval):
    """A Time interval with divisons of less than a chailek.
    
    The precision is set by the defining division.
    
    WARNING: Using two bases at once will cause errors."""
    moments_in_part = 1    
    # A class variable shared by all active instences.
    
    def __init__(self, days=0, hours=0, parts=0, moments=0, *,
     moments_in_part=None):
        if moments_in_part:
        # 0 would cause a division by zero, so rule that out too.
            FineTimeInterval.moments_in_part = moments_in_part
        self.moments = moments
        super().__init__(days=days, hours=hours, parts=parts)
    
    def reduce_parts(self):
        """Convert fractional parts to moments"""
        self.moments += (self.parts % 1) * self.moments_in_part
        self.parts = int(self.parts // 1)
        # Truncate fractional moments
        self.moments = int(self.moments // 1)
        # Carry the whole parts, then round the remaining moments.
        self.parts += self.moments // self.moments_in_part
        self.moments %= self.moments_in_part

    # Iteration function
    def __iter__(self) -> int:
        yield from [self.days, self.hours, self.parts, self.moments]

    # String function
    def __str__(self) -> str:
        return f"{self.days} {self.hours:2} {self.parts:4} {self.moments:2}"

    def __repr__(self) -> str:
        return (f"{self.__class__.__name__}({self.days}, {self.hours}, "
        f"{self.parts}, {self.moments}, moments_in_part="
        f"{self.moments_in_part})")

    def __truediv__(self, divisor):
        raise TypeError(f"Not supported for {self.__class__.__name__}")

File: ts/timeMeasures.ts
/**
 * TimeInterval: Represents a time duration with days, hours, and parts (chalakim).
 * Used for durations of months, years, etc.
 */
export class TimeInterval {
  days: number;
  hours: number;
  parts: number;

  /** Parts per hour (class variable, can be configured) */
  static parts_in_hour: number = 1;
  static readonly HOURS_IN_DAY = 24;

  /**
   * Create a TimeInterval
   * @param days - Number of days (can be fractional)
   * @param hours - Number of hours (can be fractional)
   * @param parts - Number of parts/chalakim (can be fractional)
   * @param parts_in_hour - Optional: override parts_in_hour for this class
   */
  constructor(
    days: number = 0,
    hours: number = 0,
    parts: number = 0,
    parts_in_hour?: number
  ) {
    if (parts_in_hour && parts_in_hour !== 0) {
      TimeInterval.parts_in_hour = parts_in_hour;
    }

    this.days = days;
    this.hours = hours;
    this.parts = parts;
    this.reduce();
  }

  /**
   * Reduce: normalize the time interval so each unit is within valid bounds.
   * - Fractional days become hours
   * - Fractional hours become parts
   * - Hours >= 24 become days
   * - Parts >= parts_in_hour become hours
   */
  reduce(): void {
    // Convert fractional days into hours
    this.hours += (this.days % 1) * TimeInterval.HOURS_IN_DAY;
    this.days = Math.floor(this.days);

    // Convert fractional hours into parts
    this.parts += (this.hours % 1) * TimeInterval.parts_in_hour;
    this.hours = Math.floor(this.hours);

    // Reduce parts (subclass can override to handle sub-part units)
    this.reduceParts();

    // Carry whole hours, then normalize remaining parts
    this.hours += Math.floor(this.parts / TimeInterval.parts_in_hour);
    this.parts %= TimeInterval.parts_in_hour;

    // Carry whole days, then normalize remaining hours
    this.days += Math.floor(this.hours / TimeInterval.HOURS_IN_DAY);
    this.hours %= TimeInterval.HOURS_IN_DAY;
  }

  /**
   * Reduce parts (can be overridden in subclasses for finer granularity)
   */
  protected reduceParts(): void {
    // In base class, just truncate to integer
    this.parts = Math.floor(this.parts);
  }

  /** Iterate over components [days, hours, parts] */
  *[Symbol.iterator](): Generator<number> {
    yield this.days;
    yield this.hours;
    yield this.parts;
  }

  /** Get all components as array */
  toArray(): number[] {
    return [this.days, this.hours, this.parts];
  }

  /** Hash code for use in sets/maps */
  hashCode(): number {
    let hash = 0;
    for (const val of this.toArray()) {
      hash = ((hash << 5) - hash) + val;
      hash = hash & hash; // Convert to 32bit integer
    }
    hash = ((hash << 5) - hash) + TimeInterval.parts_in_hour;
    return hash & hash;
  }

  /** String representation: "days hours parts" */
  toString(): string {
    return `${this.days} ${String(this.hours).padStart(2, ' ')} ${String(this.parts).padStart(4, ' ')}`;
  }

  /** Debug representation */
  [Symbol.for('nodejs.util.inspect.custom')](): string {
    return `TimeInterval(${this.days}, ${this.hours}, ${this.parts}, parts_in_hour=${TimeInterval.parts_in_hour})`;
  }

  // ==================== Comparison Operators ====================

  equals(other: TimeInterval | any): boolean {
    if (!(other instanceof TimeInterval)) return false;

    const thisArr = this.toArray();
    const otherArr = other.toArray();

    for (let i = 0; i < Math.max(thisArr.length, otherArr.length); i++) {
      const a = thisArr[i] ?? 0;
      const b = otherArr[i] ?? 0;
      if (a !== b) return false;
    }
    return true;
  }

  greaterThan(other: TimeInterval): boolean {
    const thisArr = this.toArray();
    const otherArr = other.toArray();

    for (let i = 0; i < Math.max(thisArr.length, otherArr.length); i++) {
      const a = thisArr[i] ?? 0;
      const b = otherArr[i] ?? 0;
      if (a > b) return true;
      if (a < b) return false;
    }
    return false;
  }

  greaterThanOrEqual(other: TimeInterval): boolean {
    return this.greaterThan(other) || this.equals(other);
  }

  lessThan(other: TimeInterval): boolean {
    const thisArr = this.toArray();
    const otherArr = other.toArray();

    for (let i = 0; i < Math.max(thisArr.length, otherArr.length); i++) {
      const a = thisArr[i] ?? 0;
      const b = otherArr[i] ?? 0;
      if (a < b) return true;
      if (a > b) return false;
    }
    return false;
  }

  lessThanOrEqual(other: TimeInterval): boolean {
    return this.lessThan(other) || this.equals(other);
  }

  // ==================== Math Operators ====================

  add(addend: TimeInterval): TimeInterval {
    const result = this.toArray();
    const otherArr = addend.toArray();

    for (let i = 0; i < Math.max(result.length, otherArr.length); i++) {
      result[i] = (result[i] ?? 0) + (otherArr[i] ?? 0);
    }

    return new TimeInterval(...result);
  }

  subtract(subtrahend: TimeInterval): TimeInterval {
    const result = this.toArray();
    const otherArr = subtrahend.toArray();

    for (let i = 0; i < Math.max(result.length, otherArr.length); i++) {
      result[i] = (result[i] ?? 0) - (otherArr[i] ?? 0);
    }

    return new TimeInterval(...result);
  }

  multiply(factor: number): TimeInterval {
    const result = this.toArray().map(x => x * factor);
    return new TimeInterval(...result);
  }

  divide(divisor: number | TimeInterval): number | TimeInterval {
    if (divisor instanceof TimeInterval) {
      // Ratio: compute as combined fractions
      const thisArr = this.toArray();
      const otherArr = divisor.toArray();
      const PARTS_IN_HOUR = TimeInterval.parts_in_hour;
      const HOURS_IN_DAY = TimeInterval.HOURS_IN_DAY;

      let numerator = 0;
      let denominator = 0;

      // Combine: days * (HOURS_IN_DAY * PARTS_IN_HOUR) + hours * PARTS_IN_HOUR + parts
      numerator = thisArr[0] * HOURS_IN_DAY * PARTS_IN_HOUR +
                  (thisArr[1] ?? 0) * PARTS_IN_HOUR +
                  (thisArr[2] ?? 0);

      denominator = otherArr[0] * HOURS_IN_DAY * PARTS_IN_HOUR +
                    (otherArr[1] ?? 0) * PARTS_IN_HOUR +
                    (otherArr[2] ?? 0);

      return denominator !== 0 ? numerator / denominator : 0;
    } else {
      // Scalar division: return FineTimeInterval
      const result = this.toArray().map(x => x / divisor);
      return new FineTimeInterval(...result, 0, divisor);
    }
  }

  negate(): TimeInterval {
    return this.multiply(-1);
  }

  absolute(): TimeInterval {
    if (this.lessThan(new TimeInterval(0))) {
      return this.negate();
    }
    return this;
  }
}

/**
 * TimeInWeek: A time of week (1-7 for Sunday-Saturday), or offset.
 * Wraps days to 1-7 (where 7 is Shabbat/day 0).
 */
export class TimeInWeek extends TimeInterval {
  /**
   * Create a TimeInWeek
   * @param days - Day of week (will be normalized to 1-7)
   * @param hours - Hours
   * @param parts - Parts
   */
  constructor(days: number = 0, hours: number = 0, parts: number = 0) {
    super(days, hours, parts);
  }

  /**
   * Override reduce to normalize day to 1-7 (7 = Shabbat)
   */
  override reduce(): void {
    super.reduce();

    // Normalize days to 1-7
    this.days = this.days % 7;
    // We want Shabbos (day 0 mod 7) to appear as 7
    if (this.days === 0) {
      this.days = 7;
    }
  }
}

/**
 * FineTimeInterval: Time interval with sub-part precision (moments).
 * Used when dividing intervals.
 */
export class FineTimeInterval extends TimeInterval {
  moments: number;

  /** Moments per part (class variable, set on division) */
  static moments_in_part: number = 1;

  /**
   * Create a FineTimeInterval
   * @param days - Days
   * @param hours - Hours
   * @param parts - Parts
   * @param moments - Sub-part moments
   * @param moments_in_part - Optional: override moments_in_part for this class
   */
  constructor(
    days: number = 0,
    hours: number = 0,
    parts: number = 0,
    moments: number = 0,
    moments_in_part?: number
  ) {
    if (moments_in_part && moments_in_part !== 0) {
      FineTimeInterval.moments_in_part = moments_in_part;
    }

    super(days, hours, parts);
    this.moments = moments;
    this.reduce();
  }

  /**
   * Override reduceParts to handle moments
   */
  protected override reduceParts(): void {
    // Convert fractional parts to moments
    this.moments += (this.parts % 1) * FineTimeInterval.moments_in_part;
    this.parts = Math.floor(this.parts);

    // Truncate fractional moments
    this.moments = Math.floor(this.moments);

    // Carry whole parts from moments
    this.parts += Math.floor(this.moments / FineTimeInterval.moments_in_part);
    this.moments %= FineTimeInterval.moments_in_part;
  }

  /** Iterate over components [days, hours, parts, moments] */
  override *[Symbol.iterator](): Generator<number> {
    yield this.days;
    yield this.hours;
    yield this.parts;
    yield this.moments;
  }

  /** Get all components as array */
  override toArray(): number[] {
    return [this.days, this.hours, this.parts, this.moments];
  }

  /** String representation */
  override toString(): string {
    return `${this.days} ${String(this.hours).padStart(2, ' ')} ${String(this.parts).padStart(4, ' ')} ${String(this.moments).padStart(2, ' ')}`;
  }

  /** Debug representation */
  override [Symbol.for('nodejs.util.inspect.custom')](): string {
    return `FineTimeInterval(${this.days}, ${this.hours}, ${this.parts}, ${this.moments}, moments_in_part=${FineTimeInterval.moments_in_part})`;
  }

  /**
   * Division not supported for FineTimeInterval
   */
  override divide(divisor: number | TimeInterval): never {
    throw new TypeError(
      `Not supported for ${this.constructor.name}`
    );
  }
}
